src/components/ColorRampDropdown.tsx

```diff
@@ -58,46 +84,89 @@ const ColorRampSelector: React.FC<ColorRampSelectorProps> = (propsInput): ReactE
+    return colorRamp.createGradientCanvas(120, theme.controls.height).toDataURL();
+  }, [props.useCategoricalPalettes, props.numCategories, props.selectedPalette]);
+
+  // Determine if we're currently using a preset palette; otherwise show the "Custom" tooltip.
+  let selectedPaletteName = "Custom";
+  for (const [, paletteData] of props.categoricalPalettes) {
+    if (arrayDeepEquals(paletteData.colors, props.selectedPalette)) {
+      selectedPaletteName = paletteData.name;
+      break;
+    }
+  }
+
+  ///////// Generate dropdown contents
+
+  const makeRampButtonList = (
+    colorRampData: ColorRampData[],
+    onClick: (rampData: ColorRampData) => void
+  ): ReactElement[] => {
     const contents: ReactElement[] = [];
-    const colorRampEntries = Array.from(props.colorRamps!.entries());
-    // Make a button for every color ramp
-    for (let i = 0; i < props.colorRamps.size; i++) {
-      // Show the name of the color ramp in the tooltip, but use its internal key for callbacks.
-      const [key, colorRampData] = colorRampEntries[i];
+    for (let i = 0; i < colorRampData.length; i++) {
       contents.push(
-        <Tooltip title={colorRampData.name} placement="right" key={key} trigger={["hover", "focus"]}>
-          <Button key={key} onClick={() => props.onChange(key, false)} rootClassName={styles.dropdownButton}>
-            <img src={colorRampData.colorRamp.createGradientCanvas(120, theme.controls.height).toDataURL()} />
+        <Tooltip title={colorRampData[i].name} placement="right" key={i} trigger={["hover", "focus"]}>
+          <Button onClick={() => onClick(colorRampData[i])} rootClassName={styles.dropdownButton}>
+            <img src={colorRampData[i].colorRamp.createGradientCanvas(120, theme.controls.height).toDataURL()} />
           </Button>
         </Tooltip>
       );
     }
     return contents;
+  };
+
+  // Memoize to avoid recalculating dropdown contents
+  const rampDropdownContents: ReactElement[] = useMemo(() => {
+    return makeRampButtonList(Array.from(props.colorRamps.values()), (rampData) => {
+      props.onChangeRamp(rampData.key, false);
+    });
   }, [props.colorRamps]);
 
+  const paletteDropdownContents: ReactElement[] = useMemo(() => {
+    const onClick = (paletteData: ColorRampData): void => {
+      const colors = props.categoricalPalettes.get(paletteData.key)?.colors;
+      if (colors) {
+        props.onChangePalette(colors);
+      }
+    };
+
+    // Generate color ramps from the palettes.
+    const paletteData = Array.from(props.categoricalPalettes.values());
+    // Append in the missing ColorRamp
+    const colorRampData = paletteData.map((data) => {
+      const visibleColors = data.colors.slice(0, Math.max(1, props.numCategories));
+      return { ...data, colorRamp: new ColorRamp(visibleColors, ColorRampType.HARD_STOP) };
+    });
+
+    return makeRampButtonList(colorRampData, onClick);
+  }, [props.categoricalPalettes, props.numCategories]);
```

Comment:
Nit: Did you consider combining paletteDropdownContents and rampDropdownContents into the same object (dropdownContents)? ...Since the component is always only in one mode or the other, based on props.

Reply:
Originally I was going to say calculating them separately saved some computation on each render, but it turns out that the way I was passing in categorical data made that untrue 😅 Did some cleanup on this file!